go/arrow/csv/reader.go

```diff
@@ -51,14 +55,20 @@ type Reader struct {
 // NewReader panics if the given schema contains fields that have types that are not
 // primitive types.
 func NewReader(r io.Reader, schema *arrow.Schema, opts ...Option) *Reader {
-	validate(schema)
-
 	rr := &Reader{r: csv.NewReader(r), schema: schema, refs: 1, chunk: 1}
 	rr.r.ReuseRecord = true
 	for _, opt := range opts {
 		opt(rr)
 	}
 
+	if rr.withHeader {
+		if err := rr.readHeader(); err != nil {
+			panic(fmt.Errorf("failed to read header %v", err))
+		}
+	}
+
+	validate(rr.schema)
```

Comment:
I would leave the schema validation where it was and only perform the "readHeader" part during the actual reading of the data.
this way, we can return the error, if any, instead of panicking.

so, I'd put the call to `readHeader` inside `Next()`, guarded with a `sync.Once`.